Add support for multiple characters in application grants

paxswill opened the pull request:
I think this should address #131. Backwards compatibility with old consumers of the api.core.info endpoint is maintained by using (in order of preference) the primary character, or the first character alphabetically (well, really whatever the db returns as the first character, I left that undefined) for the info returned by the old API.

I tried to work on the styling of the authorization window, and I think it looks _alright_, but it could do with some more work.

~~An area I'm still looking to improve on is allowing users to edit existing grants. This means users don't have to revoke and then re-authorize an app just to add (or remove) characters.~~ Done as of paxswill@8a823ff2d8c2b398c8bc7b38e09519aac951f0c8.

paxswill committed:
Allow multiple characters to be granted on authorizations

paxswill committed:
Fix bug where characters were being overwritten

paxswill committed:
Add migration script for old grants

paxswill committed:
Evaluate ACLs over all characters in a grant

paxswill committed:
Require at least one character for application grants

Also added error messages to the authorization view. The old messages were just being ignored.

Acidity commented on brave/core/api/core.py:
```diff
@@ -136,28 +136,58 @@ def info(self, token):
+            # Step 3: Match ACLs.
+            for group in Group.objects(id__in=request.service.groups):
+                if group.evaluate(token.user, character):
+                    tags.append(group.id)
```
Important distinction: Characters, not users, have groups. You should keep characters' groups separate.

Acidity commented on brave/core/api/core.py:
```diff
@@ -136,28 +136,58 @@ def info(self, token):
+
+            # Character has no keys registered.
+            if not key:
+                continue
```
How do you wish to handle when a character in the token gets removed from the user's account? You could 1) delete the token or 2) remove the character from the token. Just checking that there is a key is insufficient, as that key could be registered to a different account. (See brave/core/character/model.py function detach())

paxswill replied:
Hadn't thought of that, should be resolved in paxswill@8e6d155b6ad9be326a2d7aab89937d503fb4ad52

Acidity commented on brave/core/scripts/multiple_characters_grant_migration.py:
```diff
@@ -0,0 +1,16 @@
+        count += 1
+        if not dry_run:
+            grant.characters = [grant.character]
+            grant.character = None
```
It seems like it might be a good idea to leave the character be, especially if you wish to retain backwards compatibility of core.info.

paxswill replied:
Backwards compatibility is already maintained in core.info by supplying the same information in the same format as the old core.info calls. The only added information is a `characters` key, with an array of characters. I've backed out (as of paxswill@447286a8b324f602dbb60ee8e46848e74e8eae45) the `tags` key be a union of all of the characters' tags, and it is now limited to the same character as the rest of the old keys. Tag information for all characters (filtered by grant) is now available from the mappings in the `characters` array.

Acidity commented:
Is there a way for applications to specify how many characters they're willing to accept? Some applications (like mumble) don't have an easy way to handle multiple characters.

paxswill committed:
Separate tags by characters in api.core.info response

The `tags` key in the response should now be exactly the same as the old core.info response.

paxswill committed:
Fix lingering references to character instead of char

This is what I get for now testing before I push :/

paxswill committed:
Enforce that the grant's user owns all characters for that grant

paxswill committed:
Characters in a grant are now editable

Acidity commented on brave/core/application/controller/grant.py:
```diff
@@ -28,6 +30,49 @@ def __init__(self, grant):
+        valid, invalid = edit_grant(grant).native(kwargs)
+
+        new_characters = []
+        print grant.user.characters
```
Leftover testing code?

Acidity commented on brave/core/application/controller/grant.py:
```diff
@@ -28,6 +30,49 @@ def __init__(self, grant):
+            except EVECharacter.DoesNotExist:
+                continue
+            if character in grant.user.characters and value:
+                print 'adding', character
```
See above.

Acidity commented on brave/core/application/controller/grant.py:
```diff
@@ -28,6 +30,49 @@ def __init__(self, grant):
+                print 'adding', character
+                new_characters.append(character)
+            else:
+                print 'not adding', character, value
```
See above.

paxswill committed:
Welp, left debugging statements in there

paxswill committed:
Attempt to account for characters no longer appearing for a key

The first scenario that come to mind is a character transfer, but I don't have the PLEX on hand (now the disposable PLEX) to test it out.

paxswill committed:
Grant checking now pulls all API keys, then checks validity

Trying to account for the case where a user loses access to a character as part the API checks during authorization.

jboning commented on brave/core/api/core.py:
```diff
@@ -134,31 +134,81 @@ def info(self, token):
+                tags = char_tags
+
+            characters_info.append(character_info)
+        else:
```
Is there a reason there's an else here and break on line 166, instead of just returning above? I found that structure kind of confusing.

jboning commented on brave/core/application/controller/grant.py:
```diff
@@ -28,6 +30,44 @@ def __init__(self, grant):
+            try:
+                character = EVECharacter.objects.get(identifier=int(key))
+            except EVECharacter.DoesNotExist:
+                continue
```
would it be better to return failure instead?

paxswill replied:
Possibly, but if it failed for some reason, they could re-add the character using the grant modification UI.

jboning replied:
Yeah, definitely. I'm just worried that it won't be obvious when a character is dropped.

jboning commented on brave/core/api/core.py:
```diff
@@ -134,31 +134,81 @@ def info(self, token):
+                    character = dict(id=character.identifier, name=character.name),
+                    corporation = dict(id=character.corporation.identifier, name=character.corporation.name),
+                    alliance = dict(id=character.alliance.identifier, name=character.alliance.name) if character.alliance else None,
+                    characters = characters_info,
```
What are your thoughts on adding this here vs creating a new endpoint? Both work, I'm just curious whether there's a reason to prefer this way.

paxswill replied:
No particular reason, I just did it this way to make adding multiple character support easier to add in to the SRP app (which was what I was testing with).

jboning replied:
my kneejerk reaction was to prefer a separate endpoint... but after stewing on it for a while, I haven't come up with a good reason why, and it does make sense to provide easy access to the primary character, so I'm happy to go with this.

jboning commented:
I played with the UI some. Take a look at 94d120d

jboning commented:
@paxswill, what do you want to do with this? land as-is and tweak UI separately, or hold for UI updates here? Also, I saw mention on jabber of another idea for managing multi-character grants... do you want to pursue that design some, or...?

paxswill commented:
Yeah, thinking about it more, better management of single-character grants might be the way to go. I think it would be simpler to implement, as I think it'd just be changing the authorization/login controller. Managing which apps have grants to which characters is easier, as the grant list should already handle them. The API wouldn't change, so no backwards compatibility concerns. It'd also simplify my usage in the SRP app, as now permissions have a straightforward mapping between users and pilots.

The downside is that it complicates things for end-user. I'm percolating an idea for how to make it simpler, but it's not all there yet.
